### PR TITLE
[ASR] Separate Audio-to-Text (BPE, Char) dataset construction

### PR DESCRIPTION
**Pull request by artbataev**
Signed-off-by: Vladimir Bataev <vbataev@nvidia.com>

# What does this PR do ?

Separate Audio-to-Text Char and BPE dataset construction:
- this allows to use datasets for further mixing (e.g. in hybrid ASR-TTS models #5659)
- removes duplicated code from CTC, RNNT (Char and BPE) models

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
from nemo.collections.asr.data.audio_to_text_dataset import get_audio_to_text_bpe_dataset_from_config

dataset = get_audio_to_text_bpe_dataset_from_config(...)  # construct BPE dataset from parameters
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 1]
```diff
@@ -462,6 +464,110 @@ def get_dali_bpe_dataset(
...
+    """
+    if 'augmentor' in config:
+        augmentor = process_augmentations(config['augmentor'])
+    else:
+        augmentor = None
+
+    is_concat = config.get('is_concat', False)
+    if is_concat:
+        if 'concat_sampling' in config and config['concat_sampling'] is None:
+            logging.warning(f"Concat dataset requires `contact_sampling` but it was not provided. Config: {config}")
```
Comment (vsl9) [inline nemo/collections/asr/data/audio_to_text_dataset.py:496]: `contact_sampling` => `concat_sampling`

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 2]
```diff
@@ -462,6 +464,110 @@ def get_dali_bpe_dataset(
...
+            return None
+
+        if not 'concat_probabilities' in config:
+            logging.warning(
+                f"Concat dataset requires `contact_probabilities` list but it was not provided. Config: {config}"
```
Comment (vsl9) [inline nemo/collections/asr/data/audio_to_text_dataset.py:501]: `contact_probabilities` => `concat_probabilities`

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 3]
```diff
@@ -462,6 +464,110 @@ def get_dali_bpe_dataset(
...
+            )
+            return None
+        else:
+            if not isclose(sum(config['concat_probabilities']), 1, abs_tol=1e-6):
+                logging.warning(f"`contact_probabilities` need to sum to 1. Config: {config}")
```
Comment (vsl9) [inline nemo/collections/asr/data/audio_to_text_dataset.py:506]: `contact_probabilities` => `concat_probabilities`

Comment (artbataev) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: Fixed

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 2 above)
Comment (artbataev) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: Fixed

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 1 above)
Comment (artbataev) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: Fixed

**Review by vsl9 — APPROVED** [review]
Thanks for the refactoring!

**Review by titu1994 — CHANGES_REQUESTED** [review]
As we discussed on slack, I don't believe this PR should be merged. It is simply higher level function of already high level functions that support this. It reduces flexibility for classes that will not be able to support all features at once such as Dali, tarred, bucketing etc. 

Please keep the code separated, and duplicated as needed. We may have classes on the future which will be built that do not support all code at once
We can revisit this in the future if we see there are no classes that require special handling

**Review by titu1994 — APPROVED** [review]
Lgtm for now, we'll revisit the design in the future

**Pull request merged**